threeML/utils/histogram.py

```diff
@@ -0,0 +1,267 @@
+
+
+
+        if contents is None:
+
+            self._contents = np.zeros(len(list_of_intervals))
+
+        else:
+            assert len(list_of_intervals) == len(contents), 'contents and intervals are not the same dimension '
+
+
+            self._contents = np.array(contents)
+
+        if errors is not None:
+
+            assert len(errors) == len(contents), 'contents and errors are not the same dimension '
+
+
+            assert  is_poisson == False, 'cannot have errors and is_poisson True'
+
+
+            self._errors = np.array(errors)
+
+        else:
+
+            self._errors = None
+
+
+        if sys_errors is not None:
+
+            assert len(sys_errors) == len(contents),  'contents and errors are not the same dimension '
+
+            self._sys_errors = np.array(sys_errors)
+
+        else:
+
+            self._sys_errors = None
+
+
+
+
+        self._is_poisson = is_poisson
+
+        super(Histogram, self).__init__(list_of_intervals)
+
+        # make some assertions so that we make sure the histogram makes sense
+
+        assert self.is_contiguous(), "Histograms must have contiguous bins"
+
+        assert self.is_sorted,  "Histogram bins must be ordered"
+
+
+    def bin_entries(self, entires):
+        """
+        add the entries into the proper bin
+
+
+        :param entires: list of events
+        :return:
+        """
+
+        which_bins = np.digitize(entires, self.edges) - 1
+
+
+
+        for bin in which_bins:
+
+            try:
+
+                self._contents[bin] += 1
+
+            except(IndexError):
+                # ignore if we are outside the bins
+                pass
+
+
+
+
+    def __add__(self, other):
+
+        assert self == other, "The bins are not equal"
+
+        if self._is_poisson:
+
+            assert other.is_poisson, 'Trying to add a Poisson and non-poisson histogram together'
+
+            new_errors = None
+
+        else:
+
+            assert not other.is_poisson, 'Trying to add a Poisson and non-poisson histogram together'
+
+            if self._errors is not None:
+
+                assert other.errors is not None, "This histogram has errors, but the other does not"
+
+                new_errors = [ sqrt_sum_of_squares([e1,e2]) for e1,e2 in zip(self._errors,other.errors)]
+
+            else:
+
+                new_errors = None
+
+        if self._sys_errors is not None and other.sys_errors is not None:
+
+            new_sys_errors = [ sqrt_sum_of_squares([e1,e2]) for e1,e2 in zip(self._sys_errors,other.sys_errors)]
+
+        elif self._sys_errors is not None:
+
+            new_sys_errors = self._sys_errors
+
+        elif other.sys_errors is not None:
+
+            new_sys_errors = other.sys_errors
+
+        else:
+
+            new_sys_errors = None
+
+        new_contents = self.contents + other.contents
+
+
+        # because Hist gets inherited very deeply, when we add we will not know exactly
```

Comment:
You should call the constructor, instead of using copy, like:

new_hist = type(self)(..., new_contents, new_errors, new_sys_errors)

This of course works if the child class keeps the same calling sequence for the constructor.

Probably a better way is to rename this method _add_engine (or whatever) and have it return new_contents, new_errors, new_sys_errors.

Then you define a __add__ method like this:

```python
def __add__(self, other):
    c, e, se = self._add_engine(other)
    intervals = ...
    return type(self)(intervals, c, e, se)
```

If a child class changed the syntax of the constructor, it will need to implement a new __add__ method which will call _add_engine() and then build a new instance with the results.